### PR TITLE
[AArch64] Fold UBFMXri to UBFMWri when it's an LSR or LSL alias

### PR DESCRIPTION
**Pull request by Il-Capitano**
Using the LSR or LSL aliases of UBFM can be faster on some CPUs, so it is worth changing 64 bit UBFM instructions, that are equivalent to 32 bit LSR/LSL operations, to 32 bit variants.

This change folds the following patterns:
* If `Imms == 31` and `Immr <= Imms`:
   `UBFMXri %0, Immr, Imms`  ->  `UBFMWri %0.sub_32, Immr, Imms`
* If `Immr == Imms + 33`:
   `UBFMXri %0, Immr, Imms`  ->  `UBFMWri %0.sub_32, Immr - 32, Imms`

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp [hunk 1]
```diff
@@ -715,6 +719,45 @@ bool AArch64MIPeepholeOpt::visitFMOVDr(MachineInstr &MI) {
...
+    const TargetRegisterClass *RegClass32 =
+        TRI->getSubRegisterClass(RegClass64, AArch64::sub_32);
+
+    Register DstReg32 = MRI->createVirtualRegister(RegClass32);
+    Register DstReg64 = MI.getOperand(0).getReg();
+    Register SrcReg = MI.getOperand(1).getReg();
+
+    BuildMI(*MI.getParent(), MI, MI.getDebugLoc(), TII->get(AArch64::UBFMWri),
+            DstReg32)
+        .addReg(SrcReg, 0, AArch64::sub_32)
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp:746]: I don't think we tend to add sub-register at this point in the pipeline. It doesn't seem to cause any issues, but would it be better to add a COPY for the subreg first?

Comment (Il-Capitano) [reply llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp]: I added the extra COPY with the latest commit.

File: llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp [hunk 2]
```diff
@@ -715,6 +719,58 @@ bool AArch64MIPeepholeOpt::visitFMOVDr(MachineInstr &MI) {
...
 
+bool AArch64MIPeepholeOpt::visitUBFMXri(MachineInstr &MI) {
+  // Check if the instruction is equivalent to a 32 bit LSR or LSL alias of
+  // UBFM, and replace the UBFMXri instruction with its 32 bit variant, UBFMWri.
+  int64_t Immr = MI.getOperand(2).getImm();
+  int64_t Imms = MI.getOperand(3).getImm();
+
+  bool IsLSR = Imms == 31 && Immr <= Imms;
+  bool IsLSL = Immr == Imms + 33;
+  if (IsLSR || IsLSL) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp:730]: LLVM usually prefers to return early, lowering the indentation level.
```
if (!LSR && !IsLSL)
  return false;
```

**Review by davemgreen — APPROVED** [review]
Thanks. One last little comment but this otherwise LGTM.

File: llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp (hunk 2 above)
Comment (Il-Capitano) [reply llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp]: Thanks for catching that, fixed it now.

**Review by davemgreen — APPROVED** [review]

**Comment by Il-Capitano** [comment]
Pinging to ask for more reviews or a merge

**Comment by vitalybuka** [comment]
Not sure, but this is the only patch in the range: https://lab.llvm.org/buildbot/#/builders/55/builds/2061

**Comment by vitalybuka** [comment]
Looks like that steps is green now, not sure what is responsible for the fix and failure https://lab.llvm.org/buildbot/#/builders/55/builds/2065

**Comment by vitalybuka** [comment]
> Looks like that steps is green now, not sure what is responsible for the fix and failure https://lab.llvm.org/buildbot/#/builders/55/builds/2065

Nope, still broken.

**Comment by Il-Capitano** [comment]
I can reproduce the failure locally with 20c5432d0fb4ac852482141d89287e3fcc2d1278, which is the commit right before this got merged, so it shouldn't be this change.

**Comment by vitalybuka** [comment]
> I can reproduce the failure locally with [20c5432](https://github.com/llvm/llvm-project/commit/20c5432d0fb4ac852482141d89287e3fcc2d1278), which is the commit right before this got merged, so it shouldn't be this change.

Thanks! 
Then we need to bisect. @thurstond

**Pull request merged**